kv/memberlist/tcp_transport.go

```diff
@@ -376,24 +378,41 @@ func (t *TCPTransport) FinalAdvertiseAddr(ip string, port int) (net.IP, int, err
 				return nil, 0, fmt.Errorf("no private IP address found, and explicit IP not provided")
 			}
 
-			advertiseAddr = net.ParseIP(ip)
-			if advertiseAddr == nil {
-				return nil, 0, fmt.Errorf("failed to parse advertise address: %q", ip)
+			advertiseAddr, err = netip.ParseAddr(ip)
+			if err != nil {
+				return nil, 0, fmt.Errorf("failed to parse advertise address %q", ip)
 			}
-		} else {
+		case colonColonZero:
+			inet6Ip, err := netutil.GetFirstAddressOf(nil, t.logger, true)
+			if err != nil {
+				return nil, 0, fmt.Errorf("failed to get private inet6 address: %w", err)
+			}
+
+			addr, err := netip.ParseAddr(inet6Ip)
+			if err != nil {
+				return nil, 0, fmt.Errorf("failed to parse inet6 advertise address %q", ip)
+			}
+			advertiseAddr = addr
+		default:
 			// Use the IP that we're bound to, based on the first
 			// TCP listener, which we already ensure is there.
-			advertiseAddr = t.tcpListeners[0].Addr().(*net.TCPAddr).IP
+			addr, err := netip.ParseAddr(t.tcpListeners[0].Addr().(*net.TCPAddr).IP.String())
+			if err != nil {
+				return nil, 0, fmt.Errorf("failed to parse advertise address from tcp listener %q", ip)
+			}
+			advertiseAddr = addr
 		}
 
 		// Use the port we are bound to.
 		advertisePort = t.GetAutoBindPort()
 	}
 
-	level.Debug(t.logger).Log("msg", "FinalAdvertiseAddr", "advertiseAddr", advertiseAddr.String(), "advertisePort", advertisePort)
+	finalAddr := net.ParseIP(advertiseAddr.String())
```

Comment:
I think your right that there isn't a great reason to keep the new implementation. I was thinking that returning an error from the parsing was cleaner than `if ip4 := advertiseAddr.To4(); ip4 != nil {}` and a somewhat nicer interface to use.  I was thinking that we might adjust the signature at a later time to return `netip.Addr`, but since we don't modify the function signature here I'll put back the old implementation and we can deal with that later.

Reply:
Also I'm not sure we need the `if ip4 := advertiseAddr.To4(); ip4 != nil {}` check any more.